tfhe/src/core_crypto/fft_128_impl/math/fft/tests.rs

```diff
@@ -0,0 +1,211 @@
+            let mut fourier0_im0 = avec![0.0f64; size / 2].into_boxed_slice();
+            let mut fourier0_im1 = avec![0.0f64; size / 2].into_boxed_slice();
+
+            let mut fourier1_re0 = avec![0.0f64; size / 2].into_boxed_slice();
+            let mut fourier1_re1 = avec![0.0f64; size / 2].into_boxed_slice();
+            let mut fourier1_im0 = avec![0.0f64; size / 2].into_boxed_slice();
+            let mut fourier1_im1 = avec![0.0f64; size / 2].into_boxed_slice();
+
+            for (x, y) in izip!(poly0.as_mut().iter_mut(), poly1.as_mut().iter_mut()) {
+                *x = generator.random_uniform();
+                *y = generator.random_uniform();
+
+                *y >>= Scalar::BITS - 16;
+            }
+
+            let mut mem = GlobalPodBuffer::new(fft.backward_scratch().unwrap());
+            let mut stack = PodStack::new(&mut mem);
+
+            fft.forward_as_torus(
+                &mut fourier0_re0,
+                &mut fourier0_re1,
+                &mut fourier0_im0,
+                &mut fourier0_im1,
+                &poly0,
+            );
+            fft.forward_as_integer(
+                &mut fourier1_re0,
+                &mut fourier1_re1,
+                &mut fourier1_im0,
+                &mut fourier1_im1,
+                &poly1,
+            );
+
+            for (f0_re0, f0_re1, f0_im0, f0_im1, f1_re0, f1_re1, f1_im0, f1_im1) in izip!(
+                &mut *fourier0_re0,
+                &mut *fourier0_re1,
+                &mut *fourier0_im0,
+                &mut *fourier0_im1,
+                &*fourier1_re0,
+                &*fourier1_re1,
+                &*fourier1_im0,
+                &*fourier1_im1,
+            ) {
+                let f0_re = f128(*f0_re0, *f0_re1);
+                let f0_im = f128(*f0_im0, *f0_im1);
+                let f1_re = f128(*f1_re0, *f1_re1);
+                let f1_im = f128(*f1_im0, *f1_im1);
+
+                f128(*f0_re0, *f0_re1) = f0_re * f1_re - f0_im * f1_im;
+                f128(*f0_im0, *f0_im1) = f0_im * f1_re + f0_re * f1_im;
+            }
+
+            fft.backward_as_torus(
+                &mut convolution_from_fft,
+                &fourier0_re0,
+                &fourier0_re1,
+                &fourier0_im0,
+                &fourier0_im1,
+                stack.rb_mut(),
+            );
+            convolution_naive(
+                convolution_from_naive.as_mut(),
+                poly0.as_ref(),
+                poly1.as_ref(),
+            );
+
+            for (expected, actual) in izip!(
+                convolution_from_naive.as_ref().iter(),
+                convolution_from_fft.as_ref().iter()
+            ) {
+                let threshold = Scalar::ONE << (Scalar::BITS - 5);
+                let abs_diff = abs_diff(*expected, *actual);
+                assert!(
+                    abs_diff < threshold,
+                    "abs_diff: {abs_diff}, threshold: {threshold}",
+                );
```

Comment:
1 << (Scalar::BITS - 5) would be 2^{59} for u64 and 2^{123} for u128 right ? this bound is not tight at all right ? 🤔 or am  I misreading this ?

Reply:
question is still up for line 176 ?